### PR TITLE
[WIP] iCalendar for Scheduler

### PR DESCRIPTION
**Pull request by Joeper214**
### Features
- Upload iCalendar for each instance and save to model
- Modal and Tabbing for Manage Schedule
- Modified Viewing experience of Calendar
- Fixed error messaging

### Update process

`bundle exec rake i18n:js:export`

### CODE REVIEW

File: app/assets/javascripts/infrastructures/operation-sched-tabpane.js [hunk 1]
```diff
@@ -221,8 +225,25 @@ module.exports = Vue.extend({
       if (argument) {
         return (argument.length >= 10);
       }
-    }
+    },
+    icalendar: function (resource_id, physical_id){
+      const self = this;
```
Comment (skyarch-networks) [inline app/assets/javascripts/infrastructures/operation-sched-tabpane.js]: __[jshint]__
* 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

File: app/assets/javascripts/infrastructures/operation-sched-tabpane.js [hunk 2]
```diff
@@ -221,8 +225,25 @@ module.exports = Vue.extend({
...
+      const infra = new Infrastructure(self.infra_id);
```
Comment (skyarch-networks) [inline app/assets/javascripts/infrastructures/operation-sched-tabpane.js]: __[jshint]__
* 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

File: app/assets/javascripts/infrastructures/operation-sched-tabpane.js [hunk 3]
```diff
@@ -221,8 +225,25 @@ module.exports = Vue.extend({
...
+      self.loading = true;
+      const filename = physical_id.concat(".ics");
```
Comment (skyarch-networks) [inline app/assets/javascripts/infrastructures/operation-sched-tabpane.js]: __[jshint]__
* 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).

File: app/controllers/infrastructures_controller.rb [hunk 4]
```diff
@@ -301,6 +301,22 @@ def get_schedule
...
+  # @param [String]  physical_id
+  def icalendar
+    resource_id = params.require(:resource_id)
+    schedule = OperationDuration.find_by(resource_id: resource_id)
+    calendar = Icalendar::Calendar.new
+    calendar.add_event(schedule.to_ics)
+    calendar.publish
+
+    headers['Content-Type'] = "text/calendar; charset=UTF-8"
+    render :text => calendar.to_ical
```
Comment (skyarch-networks) [inline app/controllers/infrastructures_controller.rb]: __[rubocop]__
* Style/HashSyntax: Use the new Ruby 1.9 hash syntax. [:ref](https://github.com/bbatsov/ruby-style-guide#hash-literals)

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/a9d7a186defda05f74029193d5f43f081ccdeb84?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/assets/javascripts/operation_durations.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (skyarch-networks) [inline app/assets/javascripts/operation_durations.js]: __[jshint]__
* Expected an identifier and instead saw '#'.
* Expected an assignment or function call and instead saw an expression.
* Missing semicolon.

File: app/assets/javascripts/operation_durations.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+# All this logic will automatically be available in application.js.
```
Comment (skyarch-networks) [inline app/assets/javascripts/operation_durations.js]: __[jshint]__
* Expected an identifier and instead saw '#'.
* Expected an assignment or function call and instead saw an expression.
* Missing semicolon.

File: app/assets/javascripts/operation_durations.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
...
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (skyarch-networks) [inline app/assets/javascripts/operation_durations.js]: __[jshint]__
* Expected an identifier and instead saw '#'.
* Expected an assignment or function call and instead saw an expression.
* Missing semicolon.
* Too many errors. (75% scanned).

File: frontend/src/models/base.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+//
+// Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
+//
+// This software is released under the MIT License.
+//
+// http://opensource.org/licenses/mit-license.php
+//
+"use strict";
```
Comment (skyarch-networks) [inline frontend/src/models/base.js]: __[jshint]__
* Use the function form of "use strict".

File: frontend/src/models/base.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+Object.defineProperty(exports, "__esModule", { value: true });
```
Comment (skyarch-networks) [inline frontend/src/models/base.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/base.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    };
+    ModelBase.prototype.resolveF = function (dfd) {
+        return function (data) { return dfd.resolve(data); };
+    };
+    ModelBase.prototype.rejectF = function (dfd) {
+        return function (xhr) { return dfd.reject(xhr.responseText); };
+    };
+    return ModelBase;
+}());
+exports.default = ModelBase;
```
Comment (skyarch-networks) [inline frontend/src/models/base.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/infrastructure.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
+//
+// Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
+//
+// This software is released under the MIT License.
+//
+// http://opensource.org/licenses/mit-license.php
+//
+"use strict";
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* Use the function form of "use strict".

File: frontend/src/models/infrastructure.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+var __extends = (this && this.__extends) || (function () {
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* Possible strict violation.

File: frontend/src/models/infrastructure.js [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    var extendStatics = Object.setPrototypeOf ||
+        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* The '__proto__' property is deprecated.

File: frontend/src/models/infrastructure.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+    return function (d, b) {
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* Possible strict violation.

File: frontend/src/models/infrastructure.js [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+Object.defineProperty(exports, "__esModule", { value: true });
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/infrastructure.js [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+                instance: instance,
+                value: value
+            });
+        });
+    };
+    return Infrastructure;
+}(base_1.default));
+Infrastructure.ajax_infra = new AjaxSet.Resources('infrastructures');
+Infrastructure.ajax_resource = new AjaxSet.Resources('resources');
+exports.default = Infrastructure;
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/infrastructure.js [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+    return function (d, b) {
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+Object.defineProperty(exports, "__esModule", { value: true });
+/// <reference path="../../declares.d.ts" />
+var base_1 = require("./base");
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* 'require' is not defined.

File: frontend/src/models/infrastructure.js [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        return this.WrapAndResolveReject(function () {
+            return Infrastructure.ajax_infra.upload_calendar({
+                instance: instance,
+                value: value
+            });
+        });
+    };
+    return Infrastructure;
+}(base_1.default));
+Infrastructure.ajax_infra = new AjaxSet.Resources('infrastructures');
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* 'AjaxSet' is not defined.

File: frontend/src/models/infrastructure.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+Infrastructure.ajax_resource = new AjaxSet.Resources('resources');
```
Comment (skyarch-networks) [inline frontend/src/models/infrastructure.js]: __[jshint]__
* 'AjaxSet' is not defined.

File: frontend/src/models/operation_duration.js [hunk 20]
```diff
@@ -0,0 +1,52 @@
+//
+// Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
+//
+// This software is released under the MIT License.
+//
+// http://opensource.org/licenses/mit-license.php
+//
+"use strict";
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* Use the function form of "use strict".

File: frontend/src/models/operation_duration.js [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+var __extends = (this && this.__extends) || (function () {
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* Possible strict violation.

File: frontend/src/models/operation_duration.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    var extendStatics = Object.setPrototypeOf ||
+        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* The '__proto__' property is deprecated.

File: frontend/src/models/operation_duration.js [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+    return function (d, b) {
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* Possible strict violation.

File: frontend/src/models/operation_duration.js [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+Object.defineProperty(exports, "__esModule", { value: true });
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/operation_duration.js [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+            return Monitoring.ajax.create_host({
+                templates: templates,
+                id: _this.infra.id
+            });
+        });
+    };
+    return OperationDuration;
+}(base_1.default));
+OperationDuration.ajax = new AjaxSet.Resources('operation_durations');
+exports.default = OperationDuration;
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'exports' is not defined.

File: frontend/src/models/operation_duration.js [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+    return function (d, b) {
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+Object.defineProperty(exports, "__esModule", { value: true });
+/// <reference path="../../declares.d.ts" />
+var base_1 = require("./base");
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'require' is not defined.

File: frontend/src/models/operation_duration.js [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+            return OperationDuration.ajax.show({
+                physical_id: _this.physical_id,
+                infra_id: _this.infra.id
+            });
+        });
+    };
+    OperationDuration.prototype.create_host = function () {
+        var _this = this;
+        return this.WrapAndResolveReject(function () {
+            return Monitoring.ajax.create_host({
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'Monitoring' is not defined.

File: frontend/src/models/operation_duration.js [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+                templates: templates,
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'templates' is not defined.

File: frontend/src/models/operation_duration.js [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+                id: _this.infra.id
+            });
+        });
+    };
+    return OperationDuration;
+}(base_1.default));
+OperationDuration.ajax = new AjaxSet.Resources('operation_durations');
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.js]: __[jshint]__
* 'AjaxSet' is not defined.

File: app/controllers/infrastructures_controller.rb [hunk 30]
```diff
@@ -417,4 +353,18 @@ def infrastructure_exist
     redirect_to path, alert: msg
   end
 
+  def update_schedule(ops_exists, start_date, end_date, instance)
+    ops_exists
```
Comment (skyarch-networks) [inline app/controllers/infrastructures_controller.rb]: __[rubocop]__
* Lint/Void: Variable `ops_exists` used in void context.

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/563fdf10c0854c2f4dd2bca2d58cf77bd0bae67d?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/assets/javascripts/vue_components/radio_button.js [hunk 31]
```diff
@@ -0,0 +1,22 @@
+exports.install = function(Vue) {
+    Vue.component('radio-button', {
+        props: ['name', 'label', 'choices', 'value'],
+        template: `
```
Comment (skyarch-networks) [inline app/assets/javascripts/vue_components/radio_button.js]: __[jshint]__
* 'template literal syntax' is only available in ES6 (use 'esversion: 6').

File: frontend/src/models/operation_duration.ts [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+  show_icalendar(): JQueryPromise<any> {
+    return this.WrapAndResolveReject(() =>
+      (<any>OperationDuration.ajax).show_icalendar({
+        physical_id: this.physical_id,
+        id: this.infra_id
+      })
+    );
+  }
+
+  upload_icalendar(instance: instance[],value: string): JQueryPromise<any> {
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* missing whitespace

File: app/controllers/operation_durations_controller.rb [hunk 33]
```diff
@@ -0,0 +1,183 @@
...
+    end
+
+
+    render text: I18n.t('operation_scheduler.msg.saved'), status: 200 and return
+  end
+
+  private
+  # Use callbacks to share common setup or constraints between actions.
+  def create_schedule(start_date, end_date, instance)
+      ops = OperationDuration.create!(
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 4) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

File: app/controllers/operation_durations_controller.rb [hunk 34]
```diff
@@ -0,0 +1,183 @@
...
+    if rrules.frequency == "DAILY"
+      instance[:repeat_freq] = 1
+    else
+      if !defined? rrules.by_day
+        raise UploadCalendarError, I18n.t('operation_scheduler.msg.no_by_day')
+      end
+
+      instance[:repeat_freq] = 4
+      instance[:dates].map do |day|
+          day.second["checked"] = date_names.include?(rrules.by_day[day.second["value"].to_i]).to_s
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 4) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/c3dc233116ac5d64691f27615dcc6b1003c87e78?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: frontend/src/models/operation_duration.ts [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+import Infrastructure from './infrastructure';
+
+type instance = {
+    resource_id: string,
+    start_date: string,
+    end_date: string,
+    start_time: string,
+    end_time: string,
+    repeat_freq: string
+}
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* Missing semicolon

File: spec/controllers/operation_durations_controller_spec.rb [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+# Copyright (c) 2013-2017 SKYARCH NETWORKS INC.
+#
+# This software is released under the MIT License.
+#
+# http://opensource.org/licenses/mit-license.php
+#
+
+require_relative '../spec_helper'
+
+describe OperationDurationsController, type: :controller do
```
Comment (skyarch-networks) [inline spec/controllers/operation_durations_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [32/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/a7cb27b217d89c10f95f15ab5286e2402f97ac12?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: config/routes.rb [hunk 37]
```diff
@@ -135,6 +137,13 @@
     end
   end
 
+  resources :operation_durations, only: [:create, :show] do
```
Comment (skyarch-networks) [inline config/routes.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/operation_durations_controller.rb [hunk 38]
```diff
@@ -0,0 +1,183 @@
...
+    @resource = Resource.find_by(physical_id: physical_id)
+
+    @operation_schedule = @resource.operation_durations
+
+    if @operation_schedule.blank?
+      render text: I18n.t('operation_scheduler.msg.empty'), status: 404 and return
+    end
+
+    respond_to do |format|
+      format.json { render json: @operation_schedule.as_json(only: [:id, :start_date, :end_date],
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/operation_durations_controller.rb [hunk 39]
```diff
@@ -0,0 +1,183 @@
...
+                                                             include: [{recurring_date: {only: [:id, :repeats, :start_time, :end_time, :dates]}},
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: spec/controllers/operation_durations_controller_spec.rb [hunk 40]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (skyarch-networks) [inline spec/controllers/operation_durations_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [43/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/2d285978eb1113e786aa2fc2f6fff161c834a37e?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/controllers/operation_durations_controller.rb [hunk 41]
```diff
@@ -0,0 +1,181 @@
...
+
+
+    if rrules.frequency == "DAILY"
+      instance[:repeat_freq] = 1
+    else
+      raise UploadCalendarError, I18n.t('operation_scheduler.msg.no_by_day') if !defined? rrules.by_day
+
+      instance[:repeat_freq] = 4
+      instance[:dates].map do |day|
+          day.second["checked"] = date_names.include?(rrules.by_day[day.second["value"].to_i]).to_s
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/IndentationWidth: Use 2 (not 4) spaces for indentation. [:ref](https://github.com/bbatsov/ruby-style-guide#spaces-indentation)

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/2e735cb9d86260f3e9924ae7ee502db43264de0e?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: config/routes.rb [hunk 42]
```diff
@@ -135,6 +137,13 @@
     end
   end
 
+  resources :operation_durations, only: %i[create, show] do
```
Comment (skyarch-networks) [inline config/routes.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

File: app/controllers/operation_durations_controller.rb [hunk 43]
```diff
@@ -0,0 +1,181 @@
...
+    @resource = Resource.find_by(physical_id: physical_id)
+
+    @operation_schedule = @resource.operation_durations
+
+    if @operation_schedule.blank?
+      render text: I18n.t('operation_scheduler.msg.empty'), status: 404 and return
+    end
+
+    respond_to do |format|
+      format.json { render json: @operation_schedule.as_json(only: %i[id, start_date, end_date],
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

File: app/controllers/operation_durations_controller.rb [hunk 44]
```diff
@@ -0,0 +1,181 @@
...
+                                                             include: [{recurring_date: {only: %i[id, repeats, start_time, end_time, dates]}},
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/491bb834da4e2b380ee4532876013036e2a7151b?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: frontend/src/models/operation_duration.ts (hunk 35 above)
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* Missing semicolon

File: frontend/src/models/operation_duration.ts (hunk 32 above)
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* missing whitespace

File: config/routes.rb (hunk 42 above)
Comment (skyarch-networks) [inline config/routes.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

File: app/controllers/operation_durations_controller.rb (hunk 43 above)
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

File: app/controllers/operation_durations_controller.rb (hunk 44 above)
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Lint/PercentSymbolArray: Within `%i`/`%I`, ':' and ',' are unnecessary and may be unwanted in the resulting symbols.

File: spec/controllers/operation_durations_controller_spec.rb (hunk 40 above)
Comment (skyarch-networks) [inline spec/controllers/operation_durations_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [43/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/491bb834da4e2b380ee4532876013036e2a7151b?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: config/routes.rb (hunk 37 above)
Comment (skyarch-networks) [inline config/routes.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/664fc815bcdb580acc8a3684d6a2ba832be6de84?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

File: app/controllers/operation_durations_controller.rb [hunk 45]
```diff
@@ -0,0 +1,181 @@
...
+    @resource = Resource.find_by(physical_id: physical_id)
+
+    @operation_schedule = @resource.operation_durations
+
+    if @operation_schedule.blank?
+      render text: I18n.t('operation_scheduler.msg.empty'), status: 404 and return
+    end
+
+    respond_to do |format|
+      format.json { render json: @operation_schedule.as_json(only: [:id, :start_date, :end_date],
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: app/controllers/operation_durations_controller.rb [hunk 46]
```diff
@@ -0,0 +1,181 @@
...
+                                                             include: [{recurring_date: {only: [:id, :repeats, :start_time, :end_time, :dates]}},
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[rubocop]__
* Style/SymbolArray: Use `%i` or `%I` for an array of symbols. [:ref](https://github.com/bbatsov/ruby-style-guide#percent-i)

File: spec/controllers/operation_durations_controller_spec.rb [hunk 47]
```diff
@@ -0,0 +1,71 @@
...
```
Comment (skyarch-networks) [inline spec/controllers/operation_durations_controller_spec.rb]: __[rubocop]__
* Metrics/BlockLength: Block has too many lines. [46/25]

**Comment by skyarch-networks** [comment]
Click here if you want to create rubocop's auto-correct Pull Request
[Auto Correct by SideCI](https://sideci.com/gh/31754199/issues/pull_request/371/diff/43581eac15b3442b21c29291d0b6534dc58478b6?open_modal=1#rubocop)

See the documentation if you want to get more information about auto-correct.
[Document](https://sideci.com/ja/docs/autofix-rubocop)

**Comment by Joeper214** [comment]
Please REview and Merge PR: @root-beer

**Comment by root-beer** [comment]
I'll review now.

**Comment by root-beer** [comment]
Do you mind if I add some missing translations?

**Comment by Joeper214** [comment]
Yes sure. おねがいします。

**Comment by root-beer** [comment]
Why is the `aws-sdk` gem version old?

**Comment by Joeper214** [comment]
@root-beer . Maybe this happened when I tried to fix Gemfile conflicts. I'll try to solve this. Thanks!

File: Gemfile.lock [hunk 48]
```diff
@@ -62,10 +62,10 @@ GEM
     awsecrets (1.11.0)
       aws-sdk (~> 2)
       aws_config (~> 0.1.0)
-    awspec (0.37.7)
-      activesupport
-      aws-sdk (~> 2.2)
-      awsecrets (~> 1)
+    awspec (0.81.1)
+      activesupport (~> 4.0)
+      aws-sdk (>= 2.2, < 2.10)
```
Comment (Joeper214) [inline Gemfile.lock]: This line makes the lock.

Comment (Joeper214) [inline Gemfile.lock]: I was a bit confused too. Not really sure why this happens.

Comment (root-beer) [inline Gemfile.lock]: I see. umm...

File: frontend/src/models/operation_duration.ts [hunk 49]
```diff
@@ -0,0 +1,70 @@
...
+  show(): JQueryPromise<any> {
+    return this.WrapAndResolveReject(() =>
+      (<any>OperationDuration.ajax).show({
+        physical_id: this.physical_id,
+        id: this.infra_id
+      })
+    );
+  }
+
+  create(instance: instance[]): JQueryPromise<any> {
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* Shadowed name: 'instance'

File: frontend/src/models/operation_duration.ts [hunk 50]
```diff
@@ -0,0 +1,70 @@
...
+  show_icalendar(): JQueryPromise<any> {
+    return this.WrapAndResolveReject(() =>
+      (<any>OperationDuration.ajax).show_icalendar({
+        physical_id: this.physical_id,
+        id: this.infra_id
+      })
+    );
+  }
+
+  upload_icalendar(instance: instance[], value: string): JQueryPromise<any> {
```
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[tslint]__
* Shadowed name: 'instance'

File: frontend/src/models/operation_duration.ts (hunk 49 above)
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[TSLint]__

Shadowed name: 'instance'

File: frontend/src/models/operation_duration.ts (hunk 50 above)
Comment (skyarch-networks) [inline frontend/src/models/operation_duration.ts]: __[TSLint]__

Shadowed name: 'instance'

File: app/controllers/operation_durations_controller.rb [hunk 51]
```diff
@@ -0,0 +1,181 @@
...
+    RecurringDate.create!(
+      operation_duration_id: ops.id,
+      repeats: instance[:repeat_freq].to_i,
+      start_time:  start_date.strftime("%H:%M"),
+      end_time: end_date.strftime("%H:%M"),
+      dates: instance[:dates]
+    )
+  end
+
+  def update_schedule(ops_exists, start_date, end_date, instance)
```
Comment (skyarch-networks) [inline app/controllers/operation_durations_controller.rb]: __[Rails Best Practices]__

replace complex creation with factory method (recur_exits attribute_assignment_count > 2)

File: app/helpers/operation_durations_helper.rb [hunk 52]
```diff
@@ -0,0 +1,2 @@
+module OperationDurationsHelper
```
Comment (skyarch-networks) [inline app/helpers/operation_durations_helper.rb]: __[Rails Best Practices]__

remove empty helpers

File: config/routes.rb [hunk 53]
```diff
@@ -138,6 +138,13 @@
     end
   end
 
+  resources :operation_durations, only: [:create, :show] do
```
Comment (skyarch-networks) [inline config/routes.rb]: __[Rails Best Practices]__

restrict auto-generated routes operation_durations (only: [])

**Comment by skyarch-networks** [comment]
RuboCop detected 11 issues on d054ad953421946d8103588d86666bd83cf12d9c
Visit https://sideci.com/gh/31754199/pull_requests/371 to review the issues.

**Pull request closed without merging**